### ISSUE TITLE
Cannot load plugins on windows.

### ISSUE THREAD

**Issue by Naddiseo** [babel/babel#1361]
I'm using a plugin through babel-loader, and the only way to pass in a custom plugin is via direct path. So I have something like:
```javascript
var loaderOpts = {
    plugins: [
        path.join(__dirname, 'my_plugin.js')
    ]
};
```
On linux this works fine, but on windows, the absolute path becomes "c:\path\to\project\my_plugin.js", on which babel gets confused because of the colon in the path name.

**Comment by jquense** [comment]
Hi @Naddiseo, 

If you need to load a plugin that is local: i.e. not a module in `node_modules`. You can do by passing in the Transformer instance itself instead of the string.

```js
var loaderOpts = {
    plugins: [
        require(path.join(__dirname, 'my_plugin.js')),
    ]
};
```

if you need to change the position of the plugin you can use the format: 
```js
var loaderOpts = {
    plugins: [
        { transformer: require(path.join(__dirname, 'my_plugin.js')), position: 'after' }
    ]
};
```

**Comment by Naddiseo** [comment]
Hi @jquense, unfortunately, neither of those work. I get "didn't export a default Transformer instance"
```
Error: ModuleBuildError: Module build failed: TypeError: Plugin {"metadata":{},"handlers":{},"opts":{},"key":"my_plugin"} didn't export a default Transformer instance
```

```javascript
// my_plugin.js
var Transformer = require('babel-core').Transformer;
module.exports = new Transformer('my_plugin', {});
```

**Comment by bendrucker** [comment]
Absolute paths definitely won't work on Windows:

https://github.com/babel/babel/blob/470ebf3a46838bb782417eea4b012fd01e658755/src/babel/transformation/file/index.js#L255

You can pretty easily check on the conditions for this:

https://github.com/babel/babel/blob/470ebf3a46838bb782417eea4b012fd01e658755/src/babel/transformation/file/index.js#L285

**Comment by jquense** [comment]
@Naddiseo they definitely work I use that pattern all the time, the problem is probably that you are using two different instances of `babel`/`babel-core` one to create the transformer and the one babel is using so it is failing an `instanceof`  check.

The most recent version of babel seems to forgo the instanceof check, as @bendrucker points out in the link, for something that works 'accross  frames" so perhaps you may need to upgrade/

**Comment by monsanto** [comment]
A few solutions to this problem:

- Back out of the split if the position is not "before" or "after". Would probably lead to confusing error messages.
- Explicitly check for `/^\w:\\/` (aka a windows drive). Doesn't work for other paths that may have a colon.
- Split only on the last colon, and require that full Windows paths explicitly include a `:before` at the end.

A combination of two and three is probably the most robust; split only the last colon, and if that doesn't work, check if there is a drive in the path to give a nice error message + warning to put `:before` at the end.

**Comment by Naddiseo** [comment]
@bendrucker, exactly why I reported this as a bug; it should handle windows paths.

@jquense, that's probably the case, it's not really document which of the packages I need to install to get anything done. I have babel, babel-core, and babel-runtime in my `package.json`. I've installed the latest of them all and am still running into the problem.

@monsanto, I was thinking the third solution is probably the best.

**Comment by jquense** [comment]
If we are addressing this in babel then the _best_ solution is jsut to handle the paths correctly

```js
taken mostly from joyent/node/lib/path.js
var splitDeviceRe = /^([a-zA-Z]:|[\\\/]{2}[^\\\/]+[\\\/]+[^\\\/]+)?([\\\/])?([\s\S]*?)$/;

if ( windows) {
  var result = splitDeviceRe.exec(path),
       device = result[1] || '',
       isUnc = !!device && device[1] !== ':';
  
  // if unc use the first comma if not use the second
}
```

**Comment by sebmck** [comment]
@monsanto 

> Doesn't work for other paths that may have a colon.

A colon isn't a valid filename character in every file system I tried (NTFS, EXT3, HFS). Seems pretty safe to just check if `position[0] === "/"` and cleaning it up.

**Comment by monsanto** [comment]
@sebmck I don't have an EXT3 device handy but it works on EXT4 and BTRFS

**Comment by sebmck** [comment]
Weird, it's probably rare enough not to worry about it.

**Comment by Naddiseo** [comment]
An alternate solution would be not to use a colon to separate the path from the meta data. A semi colon or question mark a la URIs could also work.

**Comment by sebmck** [comment]
It's too late to make that change.

**Comment by AluisioASG** [comment]
I vote for no. 2.  @sebmck is right on it being rare; there aren't even provisions for escaping double-colons in `$PATH`, for example.
@jquense I think that regex's overkill.

**Comment by jquense** [comment]
@AluisioASG you'd think so but that's what node and `isAbsolute` use. Part of the issue with just checking `path[0] === '\\'` is that this is a valid windows path: `\\?\c:\\root\whatever`

**Comment by Naddiseo** [comment]
For an anecdotal data point, it's not going to be rare for people in my office; I'm the only one using *nix, everyone else uses Windows, so this is going to be a daily occurrence for us. Our current work-around is for me to build the files, then commit them to git. 

This issue is triggered by local plugins on windows with babel-loader. With babel-loader, the options are passed to babel via a JSON string, thus you have to pass absolute paths and cannot `require()` the plugin.

**Comment by monsanto** [comment]
I think it was meant that colons in the path other than the drive separator are rare

**Comment by sebmck** [comment]
@Naddiseo Nobody said Windows was rare. It was that the occurence of a colon in a filename (ie. not a  drive selector like `C:`) is rare.

**Comment by RReverser** [comment]
@sebmck As a Windows user, I'd disagree. Whenever you use full path (and you do whenever you use `__dirname` which is pretty common), you have `:` in your path.

**Comment by sebmck** [comment]
@RReverser 

> Nobody said Windows was rare. It was that the occurence of a colon in a filename (ie. not a  drive selector like `C:`) is rare.

**Comment by RReverser** [comment]
@sebmck That's exact comment I replied to.

**Comment by RReverser** [comment]
Ah wait you mean colon in other places except the drive?

**Comment by sebmck** [comment]
Yeah.

> **ie. not a drive selector like `C:`**

**Comment by RReverser** [comment]
Probably I misread the issue then - so does it actually work for drives in full path?

**Comment by sebmck** [comment]
Not currently because of [this](https://github.com/babel/babel/blob/470ebf3a46838bb782417eea4b012fd01e658755/src/babel/transformation/file/index.js#L255) line.

**Comment by RReverser** [comment]
I guess explicit regex is the best solution then. I.e.:

```javascript
[, name, , position = "before"] = name.match(/^(.*)(:(before|after))?$/)
```

**Comment by sebmck** [comment]
Fixed as of 5.2.0, thanks!

**Comment by Naddiseo** [comment]
@sebmck thanks!
